packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -104,73 +108,231 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
 
-    function deposit() external payable nonReentrant {
-        uint256 newDeposits = depositsOf[msg.sender] + msg.value;
-        require(msg.sender != TRUSTED_VALIDATOR, "ValidatorPool: trusted validator cannot deposit");
+    /**
+     * @notice Deposit ETH to be used as bond.
+     */
+    function deposit() external payable {
+        _increaseBalance(msg.sender, msg.value);
+    }
+
+    /**
+     * @notice Withdraw a given amount.
+     */
+    function withdraw(uint256 _amount) external nonReentrant {
+        _decreaseBalance(msg.sender, _amount);
+        (bool success, ) = payable(msg.sender).call{ value: _amount }("");
+        require(success);
+    }
+
+    /**
+     * @notice Bond asset corresponding to the given output index. This function is called when submitting output.
+     *         Note that the amount to bond should be at least the minimum bond amount,
+     *         and the output root must also be given for later output root comparison.
+     *
+     * @param _outputIndex Index of the L2 checkpoint output.
+     * @param _amount      Amount of bond.
+     * @param _expiresAt   The Expiration timestamp of bond.
+     */
+    function createBond(
+        uint256 _outputIndex,
+        uint128 _amount,
+        uint128 _expiresAt
+    ) external {
+        require(msg.sender == address(L2_ORACLE), "ValidatorPool: sender is not L2OutputOracle");
+        require(_amount >= MIN_BOND_AMOUNT, "ValidatorPool: the bond amount is too small");
+        require(
+            _expiresAt > block.timestamp,
+            "ValidatorPool: the expiration time should be future"
+        );
+
+        Types.Bond storage bond = bonds[_outputIndex];
         require(
-            newDeposits >= MIN_DEPOSIT_AMOUNT,
-            "ValidatorPool: the total deposit is less than minimum amount"
+            bond.expiresAt == 0,
+            "ValidatorPool: bond of the given output index already exists"
         );
 
-        depositsOf[msg.sender] = newDeposits;
+        // Unbond someone's bond if available.
+        tryUnbond();
+
+        address submitter = L2_ORACLE.getSubmitter(_outputIndex);
+        _decreaseBalance(submitter, _amount);
+
+        bond.amount = _amount;
+        bond.expiresAt = _expiresAt;
+
+        emit Bonded(submitter, _outputIndex, _amount, _expiresAt);
+    }
+
+    /**
+     * @notice Increases the bond amount corresponding to the given output index.
+     *         This function is called when creating challenge.
+     *
+     * @param _challenger  Address of the challenger.
+     * @param _outputIndex Index of the L2 checkpoint output.
+     */
+    function increaseBond(address _challenger, uint256 _outputIndex) external {
+        Types.Bond storage bond = bonds[_outputIndex];
+        require(bond.expiresAt > 0, "ValidatorPool: the bond does not exist");
+
+        uint128 bonded = bond.amount;
+        _decreaseBalance(_challenger, bonded);
+        bond.amount = bonded * 2;
```

Comment:
Please use bit operation. https://ethereum.org/en/developers/docs/evm/opcodes/ `SHL` is cheaper than `MUL`.

```suggestion
        bond.amount = bonded << 1;
```